src/controllers/auth.js

```diff
@@ -0,0 +1,397 @@
+      .then(() => {
+        if (emailNew) {
+          emailService.sendEmailChanged(email, user.email);
+          emailService.sendActivationLink(email, activationToken);
+        }
+
+        tokenService.remove(userId);
+
+        res.send(
+          {
+            user: usersService.normalize(
+              {
+                ...userData,
+                id: userId,
+              }
+            ),
+            accessToken: null,
+          }
+        );
+      })
+      .catch(error => {
+        throw ApiError.UnexpectedError(error);
+      });
+  } catch (error) {
+    throw ApiError.UnexpectedError(error);
+  }
+}
+
+async function remove(req, res, next) {
+  const userId = req.params.userId;
+  const user = await usersService.getById(userId);
+
+  const {
+    password,
+  } = req.body;
+
+  const errors = {
+    password: validateUserPassword(password, user),
+  };
+
+  if (errors.password) {
+    throw ApiError.BadRequest('Validation error', errors);
+  }
+
+  try {
+    usersService
+      .removeById(userId)
+      .then(() => {
+        tokenService.remove(userId);
+        emailService.sendAccountRemoved(user.email);
+
+        res.send({
+          user: null, accessToken: null,
+        });
+      })
+      .catch(error => {
+        throw ApiError.UnexpectedError(error);
+      });
+  } catch (error) {
+    throw ApiError.UnexpectedError(error);
+  }
+}
+
+async function reset(req, res, next) {
+  const {
+    email,
+  } = req.body;
+
+  const errors = {
+    email: validateEmail(email),
+  };
+
+  if (errors.email) {
+    throw ApiError.BadRequest('Validation error', errors);
+  }
+
+  const user = await usersService.getOneByField({ email });
+
+  if (!user) {
+    throw ApiError.BadRequest('User not found');
+  }
+
+  if (user.resetToken) {
+    if (jwtService.validateResetToken(user.resetToken)) {
+      throw ApiError.BadRequest(
+        'Request token has been already sent to your email'
+      );
+    }
+  }
+
+  try {
+    const userData = usersService.normalize(user);
+    const resetToken = jwtService.generateResetToken(userData);
+
+    user.resetToken = resetToken;
+
+    user.save();
+
+    emailService.sendResetLink(email, resetToken);
+
+    res.sendStatus(200);
+  } catch (error) {
+    throw ApiError.UnexpectedError(error);
+  }
+}
+
+async function setPassword(req, res, next) {
+  const {
+    password,
+    resetToken,
+  } = req.body;
+
+  if (!resetToken) {
+    throw ApiError.BadRequest('Reset token is required');
+  }
+
+  const errors = {
+    password: validatePassword(password),
+  };
+
+  if (errors.password) {
+    throw ApiError.BadRequest('Validation error', errors);
+  }
+
+  const user = await usersService.getOneByField({ resetToken });
+
+  if (!user) {
+    throw ApiError.BadRequest('User not found or token has expired');
+  }
+
+  try {
+    const hashedPassword = crypto
+      .pbkdf2Sync(password, user.salt, 310000, 32, 'sha256')
+      .toString('hex');
```

Comment:
i think this hashing logic is reusable, u can create separate function for this and then reuse, where it's needed